odl/operator/operator.py

```diff
@@ -95,57 +80,279 @@ def _default_apply(self, x, out, *args, **kwargs):
+
+    The ``_call()`` method of `Operator` is allowed to have the
+    following signatures:
+
+    Python 2 and 3:
+        - ``_call(self, x)``
+        - ``_call(self, vec, out)``
+        - ``_call(self, x, out=None)``
+
+    Python 3 only:
+        - ``_call(self, x, *, out=None)`` (``out`` as keyword-only
+          argument)
+
+    For disambiguation, the instance name (the first argument) **must**
+    be 'self'.
+
+    The name of the ``out`` argument **must** be 'out', the second
+    argument may have any name.
+
+    Additional variable ``**kwargs`` and keyword-only arguments
+    (Python 3 only) are also allowed.
+
+    Not allowed:
+        - ``_call(self)`` -- No arguments except instance:
+        - ``_call(x)`` -- 'self' missing, i.e. ``@staticmethod``
+        - ``_call(cls, x)``  -- 'self' missing, i.e. ``@classmethod``
+        - ``_call(self, out, x)`` -- ``out`` as second argument
+        - ``_call(self, *x)`` -- Variable arguments
+        - ``_call(self, x, y, out=None)`` -- more positional arguments
+        - ``_call(self, x, out=False)`` -- default other than `None` for
+          ``out``
+
+    In particular, static or class methods are not allowed.
+
+    Parameters
+    ----------
+    cls : `class`, optional
+        The ``_call()`` method of this class is checked. If omitted,
+        provide ``unbound_call`` instead to check directly.
+    bound_call: `callable`, optional
+        Check this bound method instead of ``cls``
+    unbound_call: `callable`, optional
+        Check this unbound function instead of ``cls``
+    attr : `string`, optional
+        Check this attribute instead of ``_call``, e.g. ``__call__``
 
-        return super().__new__(mcs, name, bases, attrs)
+    Returns
+    -------
+    has_out : `bool`
+        Whether the call has an ``out`` argument
+    out_is_optional : `bool`
+        Whether the ``out`` argument is optional
+    spec : `inspect.ArgSpec` or `inspect.FullArgSpec`
+        Argument specification of the checked call function
+
+    Raises
+    ------
+    ValueError
+        if the signature of the function is malformed
+    """
+    import inspect
+    import sys
+
+    py3 = (sys.version_info.major > 2)
+
+    py2_specs = ('_call(self, x[, **kwargs])',
+                 '_call(self, x, out[, **kwargs])',
+                 '_call(self, x, out=None[, **kwargs])')
+
+    py3only_specs = ('_call(self, x, *, out=None[, **kwargs])',)
+
+    spec_msg = "\nPossible signatures are ('[, **kwargs]' means optional):\n\n"
+    spec_msg += '\n'.join(py2_specs)
+    if py3:
+        spec_msg += '\n' + '\n'.join(py3only_specs)
+    spec_msg += '\n\nStatic or class methods are not allowed.'
+
+    if sum(arg is not None for arg in (cls, bound_call, unbound_call)) != 1:
+        raise ValueError('Exactly one object to check must be given.')
+
+    if cls is not None:
+        # Get the actual implementation, including ancestors
```

Comment:
And now I remember the reason: with the `getattr` version, the instance checks after it will always evaluate to `False`:
```python
>>> class A(object):
...     @classmethod
...     def class_method(cls):
...         pass
...
...     @staticmethod
...     def static_method():
...         pass
    
>>> class_method_getattr = getattr(A, 'class_method')
>>> class_method_dict = A.__dict__.get('class_method')
>>> isinstance(class_method_getattr, classmethod)
False
>>> isinstance(class_method_dict, classmethod)
True
>>> static_method_getattr = getattr(A, 'static_method')
>>> static_method_dict = A.__dict__.get('static_method')
>>> isinstance(static_method_getattr, staticmethod)
False
>>> isinstance(static_method_dict, staticmethod)
True
```
So only by access to the raw functions, I can properly check if they are static / class or not. Maybe I should write a test for this.